modepy/matrices.py

```diff
@@ -243,42 +243,76 @@ def mass_matrix(basis, nodes):
 
 
 class _FaceMap:
+    def __init__(self, face_vertices):
+        vol_dim = face_vertices.shape[0]
+
+        self.origin = face_vertices[:, 0].reshape(-1, 1)
+        self.span = face_vertices[:, 1:vol_dim] - self.origin
+        self.face_dim = vol_dim - 1
+
+    def __call__(self, points):
+        return self.origin + np.einsum("ad,dn->an", self.span, points*0.5 + 0.5)
+
+
+class _SimplexFaceMap(_FaceMap):
     def __init__(self, face_vertices):
         """
         :arg face_vertices: an array of shape ``[dim, npts]``, where *npts*
-            should equal *dim*.
+            should equal ``dim``.
         """
         vol_dim, npts = face_vertices.shape
         if npts != vol_dim:
-            raise ValueError("face_vertices has wrong shape")
+            raise ValueError("'face_vertices' has wrong shape")
 
-        self.origin = face_vertices[:, 0]
-        self.span = face_vertices[:, 1:] - self.origin[:, np.newaxis]
+        super().__init__(face_vertices)
 
-        self.face_dim = vol_dim - 1
 
-    def __call__(self, points):
-        return (self.origin[:, np.newaxis]
-                + np.einsum("ad,dn->an", self.span, points*0.5 + 0.5))
+class _HypercubeFaceMap(_FaceMap):
+    def __init__(self, face_vertices):
+        """
+        :arg face_vertices: an array of shape ``[dim, npts]``, where *npts*
+            should equal ``2**(dim - 1)``.
+        """
+        vol_dim, npts = face_vertices.shape
+        if npts != 2**(vol_dim-1):
+            raise ValueError("'face_vertices' has wrong shape")
+
+        super().__init__(face_vertices)
 
 
-def modal_face_mass_matrix(trial_basis, order, face_vertices, test_basis=None):
+def modal_face_mass_matrix(trial_basis, order, face_vertices,
+        test_basis=None, domain=None):
     """
-    :arg face_vertices: an array of shape ``[dim, npts]``, where *npts*
-        should equal *dim*.
+    :arg face_vertices: an array of shape ``[dim, npts]``.
+    :arg domain: identifier for the reference element, can be one of
+        `"simplex"` or `"hypercube"`.
 
     .. versionadded :: 2016.1
+
+    .. versionchanged:: 2020.5
+
+        Added *domain* parameter and support for :math:`[-1, 1]^d` domains.
     """
 
     if test_basis is None:
         test_basis = trial_basis
 
-    fmap = _FaceMap(face_vertices)
+    if domain is None:
+        domain = "simplex"
+
+    if domain == "simplex":
+        from modepy.quadrature.grundmann_moeller import \
+                GrundmannMoellerSimplexQuadrature
+        fmap = _SimplexFaceMap(face_vertices)
+        quad = GrundmannMoellerSimplexQuadrature(order, fmap.face_dim)
+    elif domain == "hypercube":
+        from modepy.quadrature import LegendreGaussTensorProductQuadrature
+        fmap = _HypercubeFaceMap(face_vertices)
+        quad = LegendreGaussTensorProductQuadrature(fmap.face_dim, order)
```

Comment:
I can't help but spot that these quadrature constructors are inconsistent in arg order. Flip? (I noticed that LGTP is being added below: just in time, phew!)

Reply:
Yep! Fixed somewhere in 704e686.

Reply:
Hmm, upon closer inspection (and as you note below): It seems that the quadrature interfaces (Grundmann-Möller, Xiao-Gimbutas, Vioreanu-Rokhlin) are actually the inconsistent one here. Almost everybody (except them) use (dims, order). Any ideas on how we could switch them over to (dims, order) in a somewhat backward-compatible way?

Reply:
Nothing comes to mind. They're both positional arguments and integers, so no particular way to distinguish between them :confounded:

Reply:
Looks like we're stuck with this inconsistency. All new code should use `(dim, order)`, but I don't think there's a good reason to dredge up the quadrature getters. I've pushed in some FIXMEs at least, so that the issue doesn't propagate. a6f8939